### PR TITLE
Update logic for sending 404 and 503

### PR DESCRIPTION
**Pull request by Arun-LinkedIn**
This PR refactors OperationTracker code to use originating data center as the source of truth when sending 404 responses. 

1. If all the replicas in originating data center are up and return that blob is not present, we send 404. Config `routerOperationTrackerTerminateOnNotFoundEnabled` is used to terminate the operation immediately. 
2. If any of the originating DC replicas are not up (i.e. in leader or standby state) and the success count doesn't meet the quorum, we send 503 if there are no other errors with higher precedence (such as Authorization failure, Blob deleted, etc).

### CODE REVIEW

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 1]
```diff
@@ -419,115 +402,28 @@ public boolean hasSucceeded() {
...
 
-    // For Delete and TtlUpdate
-    // We mark the failure as due to offline replicas when we know that we couldn't find the blob in eligible replicas,
-    // and remaining replicas are enough to make the request successful, and there is atleast one offline replica in the
-    // remaining set of replicas. The offline replicas can come back up in future to make the request successful, and
-    // hence such a failure should be deemed as retryable.
-    if (hasFailedOnOriginatingDcNotFound()
-        && originatingDcTotalReplicaCount - originatingDcNotFoundCount >= replicaSuccessTarget
-        && originatingDcOfflineReplicaCount > 0) {
+    if (originatingDcNotFoundCount >= originatingDcNotFoundFailureThreshold) {
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:415]: let's add another condition before this, as long as there is a success response, we return False here.
There is some corner case here.
If we don't check all replicas in originating dc, then the threshold here is 2.
When we upload a blob to originating DC, assume there is only replica 1 and 2 get the blob and 3 didn't get it. And when we try to delete this blob, 2 become bootstrap. Now we have 1 as leader, 2 as bootstrap and 3 as standby. We have 2 replicas as leader or standby so the threshold is not -1. However we will get not found from 2 and 3. This way we would return NOT_FOUND,even when we get success from 1.

Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:415]: or we should always check all the replicas in originating DC but this way, we would return more 503 than we should.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 2]
```diff
@@ -419,115 +402,28 @@ public boolean hasSucceeded() {
...
-                + "originatingDcEligibleReplicaCount {} {} "
-                + "replicasByState = {}",
-            routerOperation, partitionId, crossColoEnabled, diskDownCount, totalNotFoundCount, totalReplicaCount,
-            replicaSuccessTarget, totalOfflineReplicaCount, allReplicaCount, originatingDcEligibleReplicaCount,
-            getBlobIdLog(), allDcReplicasByState);
-        routerMetrics.failedMaybeDueToTotalOfflineReplicasCount.inc();
-        return true;
-      }
-      return false;
+    if (originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_VALUE) {
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:410]: Threshold being valid doesn't mean the result should be inconclusive. There could be a case where the operation is terminated before any of the request is sent out, like quota rejection. In this case, we have no NOT_FOUND error returned from server and the threshold could be -1 as well. In this way, we should return false.

I guess it's better in this way
```
if (originatingDcNotFoundFailureThreshold != NOT_FOUND_THRESHOLD_INVALID_ALUE) {
    if (originatingDcNotFoundCount >= originatingDcNotFoundFailureThreshold) {
        return True;
    }
    return False;
} else {
    if (originatingDcNotFoundCount == availableReplicasInOrginatingDC) {
        return INCONCLUSIVE; // This means even if all the available replicas returns NOT_FOUND, we still can't say it's not FOUND.
    }
    return False; // As long as there is no replica not returning NOT_FOUND
}
```

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 3]
```diff
@@ -195,12 +192,9 @@ class SimpleOperationTracker implements OperationTracker {
         (Map<ReplicaState, List<ReplicaId>>) partitionId.getReplicaIdsByStates(EnumSet.allOf(ReplicaState.class), null);
     List<ReplicaId> eligibleReplicas;
     List<ReplicaId> offlineReplicas = new ArrayList<>();
-    totalOfflineReplicaCount =
-        getReplicasByState(null, EnumSet.of(ReplicaState.OFFLINE)).getOrDefault(ReplicaState.OFFLINE,
-            Collections.emptyList()).size();
-    allReplicas = partitionId.getReplicaIds().stream().collect(Collectors.toList());
-    allReplicaCount = allReplicas.size();
+    List<ReplicaId> allReplicas = new ArrayList<>(partitionId.getReplicaIds());
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:195]: it doesn't necessarily have 9 replicas, in video, it only has 5.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 2 above)
Comment (JingQianCloud) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:410]: Justin, I think the difference between your solution and Arun's solution is the case when "originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_ALUE".

In Arun's implementation, the INCONCLUSIVE stands for there is either some offline or bootstrapping originating replicas. Although for the offline replica, we didn't send any request but we should consider there is some "virtual offline status" from these replicas not existing in the external view. 
Because of that, in the Operation code, it's a "setOperationException(AmbryUnavailable)" to compensate the missing offline we should have but we didn't receive. It's not operationException.set(AmbryUnavailable) which will overwrite the final status.
 If there is any higher precedence error, the high precedence error still take the precedence and won't change anything. So in your case, it'll still return quota_rejected.


We don't have maybeFailedDueToOfflineReplicas any more in the checkAndMaybeComplete. So if we change to below, in the case for PutBlob, that we have 1 originating replica returns success and 8 replicas are offline, not sure what the status we will return. 
    if (originatingDcNotFoundCount == availableReplicasInOrginatingDC) {
        return INCONCLUSIVE; // This means even if all the available replicas returns NOT_FOUND, we still can't say it's not FOUND.
    }

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JingQianCloud) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:195]: Yes, just want to mention it's the ideal view.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 2 above)
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:410]: I don't think this is true. Check out the DeleteOperation::checkAndMaybeComplete method. when it's quota rejected, and not enough leader and standby in originating dc, it will return 503. But we want it to return  429.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 1 above)
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:435]: Good point. Thank you for mentioning the case. Updated the code.

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java [hunk 4]
```diff
@@ -656,6 +658,14 @@ public class RouterConfig {
   @Default("false")
   public final boolean routerRepairWithReplicateBlobEnabled;
 
+  /**
+   * If this config is set to {@code true} the operation tracker would check all replicas in originating data center
+   * before concluding that blob is not present. Else, it would check in total_replicas - put_success_target + 1 replicas.
+   */
+  @Config(ROUTER_OPERATION_TRACKER_CHECK_ALL_REPLICAS_FOR_NOT_FOUND)
```
[the remark this answers is not in the document]
Comment (Arun-LinkedIn) [reply ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java]: Sure. Updated the name to make it less confusing.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:198]: Sure. Added comments to mention it is from data node configs and is same as ideal view.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 2 above)
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:428]: Yeah, actually like Jing was mentioning, I actually set 503 `setOperationException` so that errors which have more precedence are not replaced. Please let me know if this is okay.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 5]
```diff
@@ -419,115 +402,28 @@ public boolean hasSucceeded() {
...
-                + "replicasByState = {}",
-            routerOperation, partitionId, crossColoEnabled, diskDownCount, totalNotFoundCount, totalReplicaCount,
-            replicaSuccessTarget, totalOfflineReplicaCount, allReplicaCount, originatingDcEligibleReplicaCount,
-            getBlobIdLog(), allDcReplicasByState);
-        routerMetrics.failedMaybeDueToTotalOfflineReplicasCount.inc();
-        return true;
-      }
-      return false;
+    if (originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_VALUE) {
+      // Since some originating DC replicas are down or in bootstrap state, we can't conclude if blob is missing.
```
[the remark this answers is not in the document]
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java]: Sure. Added.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 1 above)
Comment (JingQianCloud) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:435]: I think probably Justin brought a generic question which always exist, 
In the checkAndMaybeComplete function, for NOT_FOUND, if hasFailedOnNotFound is true, we set the exception to NOT_FOUND no matter what the previous responses we ever received are. 

```
if (operationTracker.hasFailedOnNotFound()) {
    operationException.set(BlobDoesNotExist))
}
```

In Justin's case, replica 1 returns success status. 
What if replica 1 returns BlobAuthorizationFailure/BlobExpired etc? We are supposed to return these high precedence errors but we return NOT_FOUND.

Probably something like

```
if (operationTracker.hasFailedOnNotFound() && no_exception_received_higher_than_AmbryUnavailable) {
    operationException.set(BlobDoesNotExist))
}
```

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 2 above)
Comment (JingQianCloud) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:428]: For me, I kind of feel it's easier to understand the code logic when we have two separated function, hasFailedOnNotFound and hasOfflineReplicas.
FailedOnNotFoundResult.INCONCLUSIVE basically means we have offline replica(or bootstrapping). We can use standalone hasOfflineReplicas() function. 
When hasOfflineReplicas is true and  hasFailedOnNotFound is false, we just fake one unavailable exception. But we honor the exception precedence and honor all the response we received.
Just want to know your opinion of it. 


```
if (operationTracker.hasFailedOnNotFound()) {
    if (originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_VALUE) {
        operationException.set(RouterErrorCode.BlobDoesNotExist));
} else if (operationTracker.hasOfflineReplica()) {
      // compensate the offline status we should have but we don't offline is not in the clustermap
       setOperationException(RouterErrorCode.AmbryUnavailable):
}
```

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 6]
```diff
@@ -349,42 +348,31 @@ class SimpleOperationTracker implements OperationTracker {
...
-      // 2. We need enough active replicas in the originating DC
-      // How many active replicas is considered as enough? We have to consider some extreme cases.
-      // When the put success target is 2, we at least need 2 replicas in active state so we know the replicas are up to
-      // date. If we have only one replica in active state, we might run into a scenario that we are getting NotFound
-      // from the other two inactive replicas and we should not ignore the active one.
-      originatingDcNotFoundFailureThreshold = originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1;
-    } else {
-      originatingDcNotFoundFailureThreshold = 0;
+    originatingDcTotalReplicaCount = (int) allReplicas.stream()
+        .filter(replicaId -> replicaId.getDataNodeId().getDatacenterName().equals(this.originatingDcName))
```
[the remark this answers is not in the document]
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:359]: Sure. Since it seems like .equals(null) would return false, `originatingDcTotalReplicaCount` would be 0. Added check that `originatingDcNotFoundFailureThreshold` should be > 0 in `hasFailedNotFound`.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 7]
```diff
@@ -576,10 +476,12 @@ public void onResponse(ReplicaId replicaId, TrackedRequestFinalState trackedRequ
         if (trackedRequestFinalState == TrackedRequestFinalState.NOT_FOUND) {
           totalNotFoundCount++;
           if (replicaId.getDataNodeId().getDatacenterName().equals(originatingDcName)) {
-            originatingDcNotFoundCount++;
+            if (originatingDcLeaderOrStandbyReplicas.contains(replicaId)) {
+              // Since bootstrap replicas could still be catching up, treat NotFound responses only from leader or
+              // standby replicas as valid.
+              originatingDcNotFoundCount++;
```
Comment (Arun-LinkedIn) [inline ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:482]: Thanks @justinlin-linkedin for pointing offline that we should increment not found only if they are from Originating DC.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 8]
```diff
@@ -344,28 +348,29 @@ public enum FailedOnNotFoundResult {TRUE, FALSE, INCONCLUSIVE}
...
-            : originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1;
-    if (getEligibleReplicas(originatingDcName, EnumSet.of(ReplicaState.STANDBY, ReplicaState.LEADER)).size()
-        < originatingDcNotFoundFailureThreshold) {
-      // If sufficient replicas are not in leader or stand by state, we cannot conclude that blob is missing. Reset the
-      // NotFoundFailureThreshold to invalid value.
-      originatingDcNotFoundFailureThreshold = NOT_FOUND_THRESHOLD_INVALID_VALUE;
+    int minNumOriginatingDcReplicasToSearch =
+        routerConfig.routerOperationTrackerCheckAllOriginatingReplicasUpForNotFound ? originatingDcTotalReplicaCount
+            : Math.max(originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1, 0);
+    if (minNumOriginatingDcReplicasToSearch > 0
```
[the remark this answers is not in the document]
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java]: I think we could have a case where originatingDC might be null. For example, we had a data center in Singapore earlier which was decommissioned. If we try to fetch a blob from that time, the originatingDC could be null. In that case, this count might be 0. This is probably very corner case though.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 9]
```diff
@@ -674,12 +556,15 @@ public boolean hasFailed() {
       return totalReplicaCount - failedCount < Math.max(totalReplicaCount - 1,
           routerConfig.routerPutSuccessTarget + disabledCount);
     } else {
-      // if there is no possible way to use the remaining replicas to meet the success target,
-      // deem the operation a failure.
-      if (replicaInPoolOrFlightCount + replicaSuccessCount < replicaSuccessTarget) {
+      // If blob is not found in originating DC, we can terminate the operation early. Adding
+      // routerConfig.routerOperationTrackerTerminateOnNotFoundEnabled as a safe guard to search replicas in all colos
+      // before terminating the operation.
+      if (routerConfig.routerOperationTrackerTerminateOnNotFoundEnabled && hasFailedOnNotFound()) {
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:562]: Do we plan to enable routerOperationTrackerTerminateOnNotFoundEnabled on delivery?

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 10]
```diff
@@ -419,115 +398,36 @@ public boolean hasSucceeded() {
...
-          "Terminating {} on {} due to disk down count and total Not_Found. CrossColoEnabled: {}, DiskDownCount: {},"
-              + "TotalNotFoundCount: {}, TotalReplicaCount: {}, replicaSuccessTarget: {}, OfflineReplicaCount: {},"
-              + "allReplicaCount: {} {} replicasByState = {}", routerOperation, partitionId, crossColoEnabled, diskDownCount,
-          totalNotFoundCount, totalReplicaCount, replicaSuccessTarget, totalOfflineReplicaCount, allReplicaCount,
-          getBlobIdLog(), allDcReplicasByState);
-      routerMetrics.failedOnTotalNotFoundCount.inc();
-      return true;
-    }
-    return false;
+  public boolean hasFailedOnUnavailability() {
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:426]: I think we need change this name "hasFailedOnUnavailability".
The operation is not necessary failed on unavailability. Just we have some replicas are un-available. 
Does "hasSomeUnavailability" or something like that less confusing?

Three originating replicas, replica 1 is leader replica and returns deleted, replica 2 is standby and returns deleted, replica 3 is bootstrap, when we enable check_all_originating_replica, this function will return true.
But we don't fail on unavailability. The error status should be deleted.

Also we may have an issue in the GetOperation.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java [hunk 11]
```diff
@@ -1007,12 +1007,16 @@ void checkAndMaybeComplete() {
...
         if (progressTracker.hasSucceeded() && !retainChunkExceptionOnSuccess) {
           chunkException = null;
-        } else if (chunkOperationTracker.maybeFailedDueToOfflineReplicas()) {
-          chunkException =
-              buildChunkException("Get Chunk failed because of offline replicas", RouterErrorCode.AmbryUnavailable);
-        } else if (chunkOperationTracker.hasFailedOnNotFound()){
-          chunkException =
-              buildChunkException("Get Chunk failed because of BlobNotFound", RouterErrorCode.BlobDoesNotExist);
+        } else {
+          if (chunkOperationTracker.hasFailedOnNotFound()) {
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1011]: GetBlobInfoOperation should have the same logic.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 12]
```diff
@@ -419,115 +398,36 @@ public boolean hasSucceeded() {
...
+    // If sufficient number of originating DC replicas are not available in leader or standby state, it means Ambry is
+    // unavailable.
+    return originatingDcName != null
+        && originatingDcLeaderOrStandbyReplicas.size() < originatingDcNotFoundFailureThreshold;
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:430]: please also check if there is any success count here.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java [hunk 13]
```diff
@@ -1007,12 +1007,16 @@ void checkAndMaybeComplete() {
...
+            // If operation failed due to blob not being present, send 404.
+            chunkException =
+                buildChunkException("Get Chunk failed because of BlobNotFound", RouterErrorCode.BlobDoesNotExist);
+          } else if (chunkOperationTracker.hasFailedOnUnavailability()) {
+            // If some originating DC replicas are down, and we can't conclude if blob is missing, send 503.
+            chunkException =
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1017]: hasFailedOnUnavailability only means there is some replica unavailable, not necessary mean we fail on unavailable.
Seems chunkException is overwriting the previous exception. It'll overwrite "deleted" or "expired" status.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java (hunk 11 above)
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1060]: Thank you for pointing it. Added the same logic.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 9 above)
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:567]: Yeah, was thinking that we could enable it.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 10 above)
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java]: Sure. Changed the name to `hasSomeUnavailability` and updated to call `setChunkException` instead of `exception = e `for unavailability error.

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java [hunk 14]
```diff
@@ -419,112 +398,38 @@ public boolean hasSucceeded() {
...
-          totalNotFoundCount, totalReplicaCount, replicaSuccessTarget, totalOfflineReplicaCount, allReplicaCount,
-          getBlobIdLog(), allDcReplicasByState);
-      routerMetrics.failedOnTotalNotFoundCount.inc();
+  public boolean hasSomeUnavailability() {
+    if (originatingDcName != null && originatingDcLeaderOrStandbyReplicas.size() < originatingDcTotalReplicaCount) {
+      logger.info("{} on {} has some replicas unavailable in originating DC {}, TotalReplicaCount: {},"
+              + "LeaderStandbyCount: {}, ReplicasByState: {}, {}", routerOperation.name(), partitionId, originatingDcName,
+          originatingDcTotalReplicaCount, originatingDcLeaderOrStandbyReplicas.size(), allDcReplicasByState,
+          getBlobIdLog());
+      routerMetrics.failedMaybeDueToUnavailableReplicasCount.inc();
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java:432]: This definition of this metrics is different from before now.

**Review by JingQianCloud — APPROVED** [review]
Can we add test cases to cover all the corner cases we discussed in the previous PRs if you haven't added them yet?
Thank you.

File: ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java (hunk 12 above)
Comment (Arun-LinkedIn) [reply ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java]: Sure. As we discussed offline, instead of checking success count, I updated `hasSomeUnavailability()` to return true if there is any replica in originating DC which is not in leader and standby state. Also added comments in java doc that this method can be used when operation has failed.
